### ISSUE TITLE
Add Inequality Comparisons to Rules

### ISSUE THREAD

**Issue by gwlester** [awslabs/aws-lambda-powertools-python#701]
**Is your feature request related to a problem? Please describe.**

It would be helpful to be able to compare if something is **GREATER_THAN** or **GREATER_THAN_OR_EQUAL** to a value, e.g. if the current timestamp is greater a particular timestamp.

**Describe the solution you'd like**

Add the following to the comparison operators:

- **GREATER_THAN**
- **GREATER_THAN_OR_EQUAL**
- **LESS_THAN**
- **LESS_THAN_OR_EQUAL**
- **NOT_EQUAL**

Alternatively, provide a documented way to add custom (i.e. user-supplied) conditions to the **mapping_by_action** set used in **_match_by_action**.

**Describe alternatives you've considered**

**Additional context**

**Comment by heitorlessa** [comment]
Hey @gwlester - Thanks a lot for another feature request ;)

I think a user supplied would be great here - it can unlock many other use cases customers might want.

cc @risenberg-cyberark @whardier

**Comment by whardier** [comment]
I agree.  I may attempt to patch that in.  It would allow for some useful pre optimization as well (re.compile)

**Comment by gwlester** [comment]
It would also be nice if the _init_ took a logger and o ly created one if None was passed.Sent from my T-Mobile 4G LTE Device
-------- Original message --------From: "Shane R. Spencer" ***@***.***> Date: 9/25/21  13:33  (GMT-06:00) To: awslabs/aws-lambda-powertools-python ***@***.***> Cc: Gerald Leter ***@***.***>, Mention ***@***.***> Subject: Re: [awslabs/aws-lambda-powertools-python] Add Inequality Comparisons to Rules (#701) 
I agree.  I may attempt to patch that in.  It would allow for some useful pre optimization as well (re.compile)

**Comment by gwlester** [comment]
It would also be nice if the **_init_** took a **logger** and only created one if None was passed.

**Comment by gwlester** [comment]
Maybe this could be combined with [feat(feature_flags): add in feature_flags RE_* actions and update tests/docs #692
](https://github.com/awslabs/aws-lambda-powertools-python/pull/692/files)?

**Comment by heitorlessa** [comment]
Let’s crate a separate one so it’s easier to know this is a thing for
customers reading Changelog/Release notes.

I want to accept a Logger instance everywhere in 2.0

**Comment by gwlester** [comment]
Ok, do you want a separate ticket to track "accept a Logger instance everywhere in 2.0"??

**Note** -- that should not need a major revision roll since if the code looks like the following it is compatiable:

```python
   def __init__ (self, logger=None):
       if logger == none:
           logger = logging.getLogger(__name__)
```

Is equivalent.

**Comment by risenberg-cyberark** [comment]
@heitorlessa  @gwlester  When I created the feature, I didnt want to add too many actions right at the begining but to start with sometihng that covers my basic needs. It's great that people are using it and we should add more actions.
The only concern i have is regarding schema validation here for condition value, should we keep it as it is today - just check that it's not empty, or perhaps enforce more validation in case it's greater than (check int/float etc.). what do you think?

**Comment by risenberg-cyberark** [comment]
and bring your own logger should definitely be supported.

**Comment by heitorlessa** [comment]
Only one separate ticket for supporting Logger in the feature flags
utility. Feature flags is also beta so I’m not worried in genera as we’ve
been waiting for customer feedback to improve it overall

**Comment by gwlester** [comment]
@heitorlessa, done see [Allow bring your own logger to Feature Flags #702](https://github.com/awslabs/aws-lambda-powertools-python/issues/702)

**Comment by gwlester** [comment]
> @heitorlessa @gwlester When I created the feature, I didnt want to add too many actions right at the beginning but to start with sometihng that covers my basic needs. It's great that people are using it and we should add more actions. The only concern i have is regarding schema validation here for condition value, should we keep it as it is today - just check that it's not empty, or perhaps enforce more validation in case it's greater than (check int/float etc.). what do you think?

I'd not add any additional checking.  An exception today causes a failed condition and that is enough.

**Comment by gwlester** [comment]
I think the conditions need to be named clearer than I suggeted:

- KEY_GREATER_THAN_VALUE
- KEY_GREATER_THAN_OR_EQUAL_VALUE
- KEY_LESS_THAN_VALUE
- KEY_LESS_THAN_OR_EQUAL_VALUE
- KEY_NOT_EQUAL_VALUE

I will attempt to do a Pull Request for the above on 2021.10.01

**Comment by whardier** [comment]
I personally don't like how the exception is handled in the case where matching the value of False (which results in True) would match an exception.  Not sure if this is intended to allow for exceptions to act this way by design and not disrupt the process.

**Comment by heitorlessa** [comment]
> @heitorlessa, done see [Allow bring your own logger to Feature Flags #702](https://github.com/awslabs/aws-lambda-powertools-python/issues/702)

Thanks a lot for that - I've just made the necessary changes and merged it.

**Comment by gwlester** [comment]
@heitorlessa, will this one make next week's release?

**Comment by heitorlessa** [comment]
If anyone could help getting the tests and merge conflicts in place by Monday EOD, yep it will ;-)

**Comment by gwlester** [comment]
Tests and conflicts are done.

**Comment by gwlester** [comment]
@heitorlessa -- I think this should have the label: **status/staged-next-release**

**Comment by heitorlessa** [comment]
Yes! Added earlier but was offline - all good to be released either
tomorrow or Wed the latest

**Comment by heitorlessa** [comment]
This is now available as part of today's 1.21.0 release -- Thank you for the contribution, credits given at the top ;)

https://github.com/awslabs/aws-lambda-powertools-python/releases/tag/v1.21.0
